tls/s2n_fingerprint.c

```diff
@@ -0,0 +1,331 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tls/s2n_fingerprint.h"
+
+#include "crypto/s2n_fips.h"
+#include "crypto/s2n_hash.h"
+#include "stuffer/s2n_stuffer.h"
+#include "tls/extensions/s2n_extension_list.h"
+#include "tls/s2n_client_hello.h"
+#include "tls/s2n_crypto_constants.h"
+#include "utils/s2n_blob.h"
+#include "utils/s2n_result.h"
+#include "utils/s2n_safety.h"
+
+#define S2N_JA3_FIELD_DIV ','
+#define S2N_JA3_LIST_DIV  '-'
+
+/* UINT16_MAX == 65535 */
+#define S2N_UINT16_STR_MAX_SIZE 5
+
+/* clang-format off */
+const bool s2n_grease_values[] = {
+        [0x0A] = true, [0x1A] = true, [0x2A] = true, [0x3A] = true,
+        [0x4A] = true, [0x5A] = true, [0x6A] = true, [0x7A] = true,
+        [0x8A] = true, [0x9A] = true, [0xAA] = true, [0xBA] = true,
+        [0xCA] = true, [0xDA] = true, [0xEA] = true, [0xFA] = true,
+};
```

Comment:
Do we really need a table for this? https://godbolt.org/z/4dcro7MK3

You should be able to just

```c
ENSURE_EQ((byte1 | 0xF0), 0xFA);
```

Reply:
https://www.rfc-editor.org/rfc/rfc8701.html
> These values were allocated sparsely to discourage server implementations from conditioning on them.

You're messing up the GREASE strategy Cameron, they're supposed to be hard to match on :)